File: frontend/src/app/contexts/analysis/domain/weight-analysis.ts
import { type CalendarDate, calendarDate, DateRange } from '../../../shared/api/shared.calendar';

export interface WeightAnalysisCreation {
  readonly timelineStartDate: string;
  readonly range: {
    readonly startDate: string;
    readonly endDate: string;
  };
  readonly weightMeasurements: readonly {
    readonly date: string;
    readonly dayNumber: number;
    readonly weightInKg: number;
  }[];
}

export class WeightAnalysisMeasurement {
  private constructor(
    readonly calendarDate: CalendarDate,
    readonly dayNumber: number,
    private readonly weightInKg: number,
  ) {}

  static create(
    calendarDate: CalendarDate,
    dayNumber: number,
    weightInKg: number,
  ): WeightAnalysisMeasurement {
    if (!Number.isSafeInteger(dayNumber) || dayNumber < 1) {
      throw new RangeError('day number must be a positive safe integer');
    }

    if (!Number.isFinite(weightInKg) || weightInKg <= 0) {
      throw new RangeError('weight must be a positive finite number');
    }

    return new WeightAnalysisMeasurement(calendarDate, dayNumber, weightInKg);
  }

  weightInKilograms(): number {
    return this.weightInKg;
  }
}

export class WeightAnalysis {
  private constructor(
    readonly timelineStartDate: CalendarDate,
    readonly range: DateRange,
    readonly weightMeasurements: readonly WeightAnalysisMeasurement[],
  ) {}

  static create(creation: WeightAnalysisCreation): WeightAnalysis {
    const timelineStartDate = calendarDate(creation.timelineStartDate);
    const range = new DateRange(
      calendarDate(creation.range.startDate),
      calendarDate(creation.range.endDate),
    );
    if (timelineStartDate > range.startDate) {
      throw new RangeError('timeline start date must not be after represented range start date');
    }
    const measurements = creation.weightMeasurements.map((measurement) =>
      WeightAnalysisMeasurement.create(
        calendarDate(measurement.date),
        measurement.dayNumber,
        measurement.weightInKg,
      ),
    );

    validateMeasurements(measurements, range, timelineStartDate);

    return new WeightAnalysis(timelineStartDate, range, [...measurements]);
  }
}

function validateMeasurements(
  measurements: readonly WeightAnalysisMeasurement[],
  range: DateRange,
  timelineStartDate: CalendarDate,
): void {
  let previousDate: CalendarDate | undefined;

  for (const measurement of measurements) {
    if (!range.contains(measurement.calendarDate)) {
      throw new RangeError('weight measurement must be inside the represented range');
    }

    if (measurement.dayNumber !== dayNumberFor(timelineStartDate, measurement.calendarDate)) {
      throw new RangeError('day number must match the measurement date on the timeline');
    }

    if (previousDate !== undefined && measurement.calendarDate <= previousDate) {
      throw new RangeError('weight measurements must be ordered by unique ascending dates');
    }

    previousDate = measurement.calendarDate;
  }
}

function dayNumberFor(timelineStartDate: CalendarDate, date: CalendarDate): number {
  const millisecondsPerDay = 24 * 60 * 60 * 1000;

  return (Date.parse(date) - Date.parse(timelineStartDate)) / millisecondsPerDay + 1;
}

File: backend/src/main/java/fr/sylvainjanet/tracker/analysis/domain/WeightAnalysis.java
package fr.sylvainjanet.tracker.analysis.domain;

import fr.sylvainjanet.tracker.shared.domain.DateRange;
import fr.sylvainjanet.tracker.shared.domain.WeightMeasurement;
import java.time.LocalDate;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.Objects;

public final class WeightAnalysis {

    private final LocalDate timelineStartDate;
    private final DateRange range;
    private final List<WeightMeasurement> weightMeasurements;

    private WeightAnalysis(
            LocalDate timelineStartDate,
            DateRange range,
            List<WeightMeasurement> weightMeasurements) {
        this.timelineStartDate =
                Objects.requireNonNull(timelineStartDate, "timeline start date must not be null");
        this.range = Objects.requireNonNull(range, "represented range must not be null");
        Objects.requireNonNull(weightMeasurements, "weight measurements must not be null");

        if (timelineStartDate.isAfter(range.getStartDate())) {
            throw new IllegalArgumentException(
                    "timeline start date must not be after represented range start date");
        }

        validateMeasurements(weightMeasurements, range);
        this.weightMeasurements = List.copyOf(weightMeasurements);
    }

    public static WeightAnalysis create(
            LocalDate timelineStartDate,
            DateRange range,
            List<WeightMeasurement> weightMeasurements) {
        return new WeightAnalysis(timelineStartDate, range, weightMeasurements);
    }

    public LocalDate timelineStartDate() {
        return timelineStartDate;
    }

    public DateRange range() {
        return range;
    }

    public List<WeightMeasurement> weightMeasurements() {
        return weightMeasurements;
    }

    public long dayNumberFor(LocalDate date) {
        Objects.requireNonNull(date, "date must not be null");

        if (date.isBefore(timelineStartDate)) {
            throw new IllegalArgumentException("date must not be before the timeline start date");
        }

        return ChronoUnit.DAYS.between(timelineStartDate, date) + 1;
    }

    private static void validateMeasurements(
            List<WeightMeasurement> weightMeasurements, DateRange range) {
        LocalDate previousDate = null;

        for (WeightMeasurement measurement : weightMeasurements) {
            Objects.requireNonNull(measurement, "weight measurement must not be null");

            if (!range.contains(measurement.date())) {
                throw new IllegalArgumentException(
                        "weight measurement must be inside the represented range");
            }

            if (previousDate != null && !(measurement.date().isAfter(previousDate))) {
                throw new IllegalArgumentException(
                        "weight measurements must be ordered by unique ascending dates");
            }

            previousDate = measurement.date();
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        WeightAnalysis that = (WeightAnalysis) o;
        return Objects.equals(timelineStartDate, that.timelineStartDate)
                && Objects.equals(range, that.range)
                && hasSameMeasurementInputsAs(that);
    }

    private boolean hasSameMeasurementInputsAs(WeightAnalysis that) {
        if (weightMeasurements.size() != that.weightMeasurements.size()) {
            return false;
        }

        for (int index = 0; index < weightMeasurements.size(); index++) {
            WeightMeasurement measurement = weightMeasurements.get(index);
            WeightMeasurement otherMeasurement = that.weightMeasurements.get(index);

            if (!Objects.equals(measurement.date(), otherMeasurement.date())
                    || !Objects.equals(measurement.weight(), otherMeasurement.weight())) {
                return false;
            }
        }

        return true;
    }

    @Override
    public int hashCode() {
        int result = Objects.hash(timelineStartDate, range);

        for (WeightMeasurement measurement : weightMeasurements) {
            result = 31 * result + Objects.hash(measurement.date(), measurement.weight());
        }

        return result;
    }
}
